pkg/controller/worker/machine_dependencies.go

```
@@ -1,25 +1,181 @@
+	"github.com/gardener/gardener-extension-provider-openstack/pkg/apis/openstack/helper"
+	osclient "github.com/gardener/gardener-extension-provider-openstack/pkg/openstack/client"
+	extensionsv1alpha1 "github.com/gardener/gardener/pkg/apis/extensions/v1alpha1"
+)
+
+func (w *workerDelegate) DeployMachineDependencies(ctx context.Context) error {
+	computeClient, err := w.openstackClient.Compute()
+	if err != nil {
+		return err
+	}
+
+	workerStatus, err := w.decodeWorkerProviderStatus()
+	if err != nil {
+		return err
+	}
+
+	serverGroupDepSet, err := w.reconcileServerGroups(computeClient, workerStatus.DeepCopy())
+	return w.updateMachineDependenciesStatus(ctx, workerStatus, serverGroupDepSet.extract(), err)
+}
+
+func (w *workerDelegate) reconcileServerGroups(computeClient osclient.Compute, workerStatus *api.WorkerStatus) (serverGroupDependencySet, error) {
+	serverGroupDepSet := newServerGroupDependencySet(workerStatus.ServerGroupDependencies)
+	for _, pool := range w.worker.Spec.Pools {
+		serverGroupDependencyStatus, err := w.reconcilePoolServerGroup(computeClient, pool, serverGroupDepSet)
+		if err != nil {
+			return serverGroupDepSet, fmt.Errorf("reconciling server groups failed for pool %q: %w", pool.Name, err)
+		}
+		serverGroupDepSet.upsert(serverGroupDependencyStatus)
+	}
+	return serverGroupDepSet, nil
+}
+
+func (w *workerDelegate) reconcilePoolServerGroup(computeClient osclient.Compute, pool extensionsv1alpha1.WorkerPool, set serverGroupDependencySet) (*api.ServerGroupDependency, error) {
+	poolProviderConfig, err := helper.WorkerConfigFromRawExtension(pool.ProviderConfig)
+	if err != nil {
+		return nil, err
+	}
+
+	if poolProviderConfig == nil || poolProviderConfig.ServerGroupConfig == nil || poolProviderConfig.ServerGroupConfig.Policy == "" {
+		return nil, nil
+	}
+
+	poolDep := set.getByPoolName(pool.Name)
+	if poolDep != nil {
+		serverGroup, err := computeClient.GetServerGroup(poolDep.ID)
+		if err != nil && !osclient.IsNotFoundError(err) {
```

What's the expected behaviour when a "not found" error is returned? At the moment, it's silently ignored, right? Just to double-check that one.

---

Yes, we silently ignore it. We will create a new server group and overwrite the old serverGroupID with the new in our worker's dependencies status. 

Later in the worker reconciliation we will garbage collect all serverGroups that match our "name-tag" `clusterName-workerName` but have no reference in our dependencies status.

---

OK, thanks